packages/integration/src/gerrit/core.ts

```diff
@@ -0,0 +1,158 @@
+ * with links to the url used by the url reader. These will not work unless
+ * the urls point to an actual Gitiles installation.
+ *
+ * Gitiles url:
+ * https://g.com/optional_path/{project}/+/refs/heads/{branch}/{filePath}
+ *
+ *
+ * @param url - An URL pointing to a file stored in git.
+ * @public
+ */
+
+export function parseGitilesUrl(
+  config: GerritIntegrationConfig,
+  url: string,
+): GitFile {
+  const urlPath = url.replace(config.gitilesBaseUrl!, '');
+  const parts = urlPath.split('/').filter(p => !!p);
+
+  const projectEndIndex = parts.indexOf('+');
+
+  if (projectEndIndex <= 0) {
+    throw new Error(`Unable to parse project from url: ${url}`);
+  }
+  const project = trimStart(parts.slice(0, projectEndIndex).join('/'), '/');
+
+  const branchIndex = parts.indexOf('heads');
+  if (branchIndex <= 0) {
+    throw new Error(`Unable to parse branch from url: ${url}`);
+  }
+  const branch = parts[branchIndex + 1];
+  const filePath = parts.slice(branchIndex + 2).join('/');
+
+  return {
+    branch,
+    filePath: filePath === '' ? '/' : filePath,
+    project,
+  };
+}
+
+/**
+ * Return the authentication prefix.
+ *
+ * @remarks
+ *
+ * To authenticate with a password the API url must be prefixed with "/a/".
+ * If no password is set anonymous access (without the prefix) will
+ * be used.
+ *
+ * @param config - A Gerrit provider config.
+ * @public
+ */
+export function getAuthenticationPrefix(
+  config: GerritIntegrationConfig,
+): string {
+  return config.password ? '/a/' : '/';
+}
+
+/**
+ * Return the url to fetch the contents of a file using the Gerrit API.
+ *
+ * @param url - An url pointing to a file in git.
+ * @public
+ */
+export function getGerritFileContentsApiUrl(
+  config: GerritIntegrationConfig,
+  url: string,
+) {
+  const { branch, filePath, project } = parseGitilesUrl(config, url);
+
+  return `${config.baseUrl}${getAuthenticationPrefix(
+    config,
+  )}projects/${encodeURIComponent(
+    project,
+  )}/branches/${branch}/files/${encodeURIComponent(filePath)}/content`;
+}
+
+/**
+ * Return request headers for a Gerrit provider.
+ *
+ * @param config - A Gerrit provider config
+ * @public
+ */
+export function getGerritRequestOptions(config: GerritIntegrationConfig) {
+  const headers: Record<string, string> = {};
+
+  if (!config.password) {
+    return headers;
+  }
+  const buffer = Buffer.from(`${config.username}:${config.password}`, 'utf8');
+  headers.Authorization = `Basic ${buffer.toString('base64')}`;
+  return {
+    headers,
+  };
+}
+
+/**
+ * Parse the json response from Gerrit and strip the magic prefix.
+ *
+ * @remarks
+ *
+ * To prevent against XSSI attacks the JSON response body from Gerrit starts
+ * with a magic prefix that must be stripped before it can be fed to a JSON
+ * parser.
+ *
+ * @param response - An API response.
+ * @public
+ */
+export async function parseGerritJsonResponse(
+  response: Response,
+): Promise<any> {
+  const responseBody = await response.text();
```

Comment:
Hm. Would feel prudent here to actually check that the response starts with those three specific characters. It's extremely unlikely that this would ever turn into an actual attack vector in practice, but still :)